framework/auth/tasks.py

```diff
@@ -27,3 +38,122 @@ def update_user_from_activity(user_id, login_time, cas_login=False, updates=None
         should_save = True
     if should_save:
         user.save()
+
+
+@celery_app.task()
+def update_affiliation_for_orcid_sso_users(user_id, orcid_id):
+    """This is an asynchronous task that runs during CONFIRMED ORCiD SSO logins and makes eligible
+    institution affiliations.
+    """
+    from osf.models import OSFUser
+    user = OSFUser.load(user_id)
+    if not user or not verify_user_orcid_id(user, orcid_id):
+        # This should not happen as long as this task is called at the right place at the right time.
+        error_message = f'Invalid ORCiD ID [{orcid_id}] for [{user_id}]' if user else f'User [{user_id}] Not Found'
+        logger.error(error_message)
+        sentry.log_message(error_message)
+        return
+    institution = check_institution_affiliation(orcid_id)
+    if institution:
+        logger.info(f'Eligible institution affiliation has been found for ORCiD SSO user: '
+                    f'institution=[{institution._id}], user=[{user_id}], orcid_id=[{orcid_id}]')
+        if not user.is_affiliated_with_institution(institution):
+            user.affiliated_institutions.add(institution)
+
+
+def verify_user_orcid_id(user, orcid_id):
```

Comment:
uber nit: we added a [`get_verified_external_id`](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/models/user.py#L667) function recently, so this can use that instead of surfacing the messy implementation details of the `external_identity' dictionary -- unless there's a real risk of one user having multiple verified orcids and us needing to check this specific one.

Reply:
>unless there's a real risk of one user having multiple verified orcids and us needing to check this specific one

It's not impossible, and is an edge case that would cause the functionality added in this PR to misbehave. It's usually safe to assume that any possible edge case will eventually be hit.